File: server/shared/security.js
/**
 * Security Middleware — Shared across auth and agent servers
 * Rate limiting, CORS config, Helmet headers, input sanitization
 */
const rateLimit = require('express-rate-limit');
const helmet = require('helmet');
const cors = require('cors');

/**
 * Parse allowed origins from environment variable.
 * Falls back to localhost in development.
 */
function getAllowedOrigins() {
    const envOrigins = process.env.ALLOWED_ORIGINS;
    if (envOrigins) {
        return envOrigins.split(',').map(o => o.trim());
    }
    // Default development origins
    return [
        'http://localhost:5173',
        'http://localhost:3000',
        'http://localhost:5000'
    ];
}

/**
 * Configured CORS middleware — restricts to allowed origins only.
 */
function secureCors() {
    const allowedOrigins = getAllowedOrigins();

    return cors({
        origin: function (origin, callback) {
            // Allow requests with no origin (mobile apps, curl, server-to-server)
            if (!origin) return callback(null, true);

            if (allowedOrigins.includes(origin)) {
                return callback(null, true);
            }

            // In development, allow all localhost origins
            if (process.env.NODE_ENV !== 'production' && origin.startsWith('http://localhost')) {
                return callback(null, true);
            }

            return callback(new Error(`Origin ${origin} not allowed by CORS`));
        },
        credentials: true,
        methods: ['GET', 'HEAD', 'PUT', 'PATCH', 'POST', 'DELETE'],
        allowedHeaders: ['Origin', 'X-Requested-With', 'Content-Type', 'Accept', 'Authorization'],
        maxAge: 86400 // Cache preflight for 24h
    });
}

/**
 * Security headers via Helmet — sets Content-Security-Policy, HSTS, etc.
 */
function securityHeaders() {
    return helmet({
        contentSecurityPolicy: process.env.NODE_ENV === 'production' ? undefined : false,
        crossOriginEmbedderPolicy: false // Allow iframe embedding for agent
    });
}

// ─── Rate Limiters ──────────────────────────────────────────

/**
 * General API rate limiter — 100 requests per 15 minutes per IP.
 */
function generalLimiter() {
    return rateLimit({
        windowMs: parseInt(process.env.RATE_LIMIT_WINDOW_MS || '900000', 10), // 15 min
        max: parseInt(process.env.RATE_LIMIT_MAX_REQUESTS || '100', 10),
        standardHeaders: true,
        legacyHeaders: false,
        message: {
            success: false,
            message: 'Too many requests. Please try again later.'
        }
    });
}

/**
 * Auth rate limiter — strict: 10 attempts per 15 minutes for login/signup.
 * Prevents brute-force attacks on credentials.
 */
function authLimiter() {
    return rateLimit({
        windowMs: 15 * 60 * 1000, // 15 minutes
        max: 10,
        standardHeaders: true,
        legacyHeaders: false,
        message: {
            success: false,
            message: 'Too many authentication attempts. Please wait 15 minutes.'
        },
        keyGenerator: (req) => {
            // Rate limit by email to prevent brute-force on one account
            return req.body?.email || 'unknown';
        },
        validate: { xForwardedForHeader: false }
    });
}

/**
 * AI chat rate limiter — moderate: 30 requests per minute per user.
 * Each request costs Gemini API credits.
 */
function chatLimiter() {
    return rateLimit({
        windowMs: 60 * 1000, // 1 minute
        max: 30,
        standardHeaders: true,
        legacyHeaders: false,
        message: {
            success: false,
            message: 'Too many AI requests. Please wait a moment before trying again.'
        },
        keyGenerator: (req) => {
            // Use authenticated userId if available, else let default IP key handle it
            return req.userId || req.headers['x-forwarded-for'] || 'anonymous';
        },
        validate: { xForwardedForHeader: false }
    });
}

/**
 * Basic input sanitization — strips HTML/script tags from string values.
 * Apply as middleware on routes that accept user text input.
 */
function sanitizeBody(req, res, next) {
    if (req.body && typeof req.body === 'object') {
        for (const key of Object.keys(req.body)) {
            if (typeof req.body[key] === 'string') {
                req.body[key] = req.body[key]
                    .replace(/<[^>]*>/g, '')  // Strip HTML tags
                    .trim();
            }
        }
    }
    next();
}

module.exports = {
    secureCors,
    securityHeaders,
    generalLimiter,
    authLimiter,
    chatLimiter,
    sanitizeBody,
    getAllowedOrigins
};

File: server_py/shared/security.py
"""
Security Middleware — Shared across auth and agent subsystems.
Rate limiting, CORS config, input sanitization.
"""

import re
import time
from fastapi import Request, Response
from fastapi.middleware.cors import CORSMiddleware
from slowapi import Limiter
from slowapi.util import get_remote_address
from starlette.middleware.base import BaseHTTPMiddleware

from config import get_settings

# ─── Rate Limiter Instance ───────────────────────────────────
limiter = Limiter(key_func=get_remote_address)


def get_allowed_origins() -> list[str]:
    """Parse allowed origins from environment variable."""
    settings = get_settings()
    origins = settings.allowed_origins_list
    if not origins:
        return [
            "http://localhost:5173",
            "http://localhost:3000",
            "http://localhost:5000",
        ]
    return origins


def configure_cors(app):
    """Add CORS middleware to the FastAPI app."""
    settings = get_settings()
    origins = get_allowed_origins()

    # In development, allow all localhost origins
    if not settings.is_production:
        origins.extend([
            "http://localhost:5173",
            "http://localhost:3000",
            "http://localhost:5000",
        ])
        origins = list(set(origins))

    app.add_middleware(
        CORSMiddleware,
        allow_origins=origins,
        allow_credentials=True,
        allow_methods=["GET", "HEAD", "PUT", "PATCH", "POST", "DELETE"],
        allow_headers=["Origin", "X-Requested-With", "Content-Type", "Accept", "Authorization"],
        max_age=86400,
    )


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Adds security headers equivalent to Helmet.js."""

    async def dispatch(self, request: Request, call_next) -> Response:
        response = await call_next(request)

        # Helmet-equivalent headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "SAMEORIGIN"
        response.headers["X-XSS-Protection"] = "0"
        response.headers["Strict-Transport-Security"] = "max-age=15552000; includeSubDomains"
        response.headers["Referrer-Policy"] = "no-referrer"
        response.headers["X-Download-Options"] = "noopen"
        response.headers["X-Permitted-Cross-Domain-Policies"] = "none"

        return response


def sanitize_string(value: str) -> str:
    """Strip HTML/script tags from a string value."""
    if not isinstance(value, str):
        return value
    return re.sub(r"<[^>]*>", "", value).strip()


def sanitize_body_dict(body: dict) -> dict:
    """Strip HTML from all string values in a request body dict."""
    sanitized = {}
    for key, value in body.items():
        if isinstance(value, str):
            sanitized[key] = sanitize_string(value)
        else:
            sanitized[key] = value
    return sanitized


# ─── Rate Limit Key Generators ───────────────────────────────

def auth_key_func(request: Request) -> str:
    """Rate limit auth endpoints by email to prevent brute-force."""
    # Try to extract email from body — fallback to IP
    return get_remote_address(request)


def chat_key_func(request: Request) -> str:
    """Rate limit chat by userId if authenticated, else IP."""
    user_id = getattr(request.state, "user_id", None)
    if user_id:
        return str(user_id)
    return get_remote_address(request)
